### PR TITLE
adding the notes table and versioning

### PR DESCRIPTION
**Pull request by femalves**
I would really like your input this. I wrote the model but I'm not sure if it's the best approach. Please, let me know your thoughts. 
I also made some changes to manage.py but I haven't fully reviewed those yet, but feel free to give me some pointers.

### CODE REVIEW

File: biblib/views/library_view.py [hunk 1]
```diff
@@ -152,71 +231,408 @@ def solr_update_library(library_id, solr_docs):
...
+        except Exception as error:
+            current_app.logger.warning(
+                'Library missing or solr endpoint failed: {0}'
+                .format(error)
+            )
+            return data["library_id"], None, err(MISSING_LIBRARY_ERROR)
+        
+    def is_library_public_or_has_special_token(self, library, request):
+        """
+        HTTP GET request that returns all the documents inside a given
```
Comment (kelockhart) [inline biblib/views/library_view.py]: Copy/paste error - update text

File: biblib/views/library_view.py [hunk 2]
```diff
@@ -277,156 +696,58 @@ def get(self, library):
...
-        # If they do not have access, exit
-        if not self.read_access(service_uid=service_uid,
-                                library_id=library.id):
-            current_app.logger.error(
-                'User: {0} does not have access to library: {1}. DENIED'
-                .format(service_uid, library.id)
-            )
-            return err(NO_PERMISSION_ERROR)
+        
+        current_app.logger.warning('Library: {0} is private'.format(library_id))
```
Comment (kelockhart) [inline biblib/views/library_view.py]: I know this is pre-existing code, but it's weird that it's generating a warning-level log for this - can you downgrade it to info-level?

File: biblib/tests/unit_tests/test_models.py [hunk 3]
```diff
@@ -3,10 +3,12 @@
 """
 
 import unittest
-from biblib.models import User, Library, Permissions, MutableDict
+from biblib.models import User, Library, Permissions, MutableDict, Notes
 from biblib.tests.base import TestCaseDatabase
+from sqlalchemy import exc
+import pytest
```
Comment (kelockhart) [inline biblib/tests/unit_tests/test_models.py:9]: Is this leftover from testing? I don't see it in here

File: biblib/tests/unit_tests/test_views.py [hunk 4]
```diff
@@ -1,16 +1,16 @@
 """
 Tests Views of the application
 """
-
+import json
```
Comment (kelockhart) [inline biblib/tests/unit_tests/test_views.py]: Leftover from testing?

**Review by kelockhart — CHANGES_REQUESTED** [review]
Looking good so far! There are a few minor comments, but overall it's looking pretty solid, and the refactored code is much easier to read. I know I started this review before your latest commits, so you might have addressed some of this already - if so, disregard the comments that don't apply.

File: biblib/manage.py [hunk 5]
```diff
@@ -81,50 +81,61 @@ def run(app=app, n_years=None):
...
+                session.rollback()
+
     @staticmethod
     def run(app=app, n_revisions=None):
-        """
-        Carries out the deletion of older versions
-        """
-        if not n_revisions: n_revisions = current_app.config.get('NUMBER_REVISIONS', 7)
+        if not n_revisions:
+            n_revisions = current_app.config.get('NUMBER_REVISIONS', 7)
```
[the remark this answers is not in the document]
Comment (femalves) [reply biblib/manage.py:131]: Okay. I'll keep that in mind.

File: biblib/models.py [hunk 6]
```diff
@@ -172,7 +172,52 @@ def __repr__(self):
...
+class Notes(Base): 
+    """ 
+    Notes table 
+    """
+             
+    __tablename__ = 'notes'
+    __versioned__ = {}
+    id = Column(Integer, primary_key=True)
+    content = Column(UnicodeText)
+    bibcode = Column(String(50), nullable=False)
```
[the remark this answers is not in the document]
Comment (femalves) [reply biblib/models.py:184]: Done.

**Review by tjacovich — CHANGES_REQUESTED** [review]
I think the merge with `timestamp_sort` looks good. Just had a couple of minor comments.

File: biblib/views/user_view.py [hunk 7]
```diff
@@ -151,9 +152,18 @@ def get_libraries(cls, service_uid, absolute_uid, start=0, rows=None, sort_col="
                     owner=owner
                 )
 
-                output_libraries.append(payload)
-
-            return output_libraries
+                if (permissions and main_permission in ['owner']) or not permissions:
```
Comment (kelockhart) [inline biblib/views/user_view.py:155]: Where is `permissions` set? I see `permission.permissions` but not just `permissions`, but maybe I'm missing it. Either way, I have some concerns about the `or not permissions` part being lumped in with the `'owner'` permission. Depending on how `permissions` is assigned, either stick the `or not permissions` in the `elif`, or an `else`

**Review by kelockhart — CHANGES_REQUESTED** [review]
Sorry this took me so long! I have some high level questions and suggestions, along with a couple more detailed comments. Let's chat about this during our next meeting - a synchronous review will go faster on my part.

**Review by tjacovich — APPROVED** [review]
Looks good to me. Excited to see it in action.

**Pull request merged**